Orange/widgets/data/owfeatureconstructor.py

```diff
@@ -847,10 +887,36 @@ def bind_variable(descriptor, env):
                    if name in variables]
 
     values = {}
+    cast = None
+    nan = float("nan")
+
     if isinstance(descriptor, DiscreteDescriptor):
-        values = [sanitized_name(v) for v in descriptor.values]
-        values = {name: i for i, name in enumerate(values)}
-    return descriptor, FeatureFunc(descriptor.expression, source_vars, values)
+        if not descriptor.values:
+            str_func = FeatureFunc(descriptor.expression, source_vars)
+            values = sorted({str(x) for x in str_func(data)})
+            values = {name: i for i, name in enumerate(values)}
+            descriptor = descriptor._replace(values=values)
+
+            def cast(x):  # pylint: disable=function-redefined
+                return values.get(x, nan)
+
+        else:
+            values = [sanitized_name(v) for v in descriptor.values]
+            values = {name: i for i, name in enumerate(values)}
+
+    if isinstance(descriptor, DateTimeDescriptor):
+        parse = Orange.data.TimeVariable("_").parse
+
+        def cast(e):  # pylint: disable=function-redefined
+            if isinstance(e, float):
+                return e
+            try:
+                return parse(e)
+            except:  # pylint: disable=bare-except
+                return nan
```

Comment:
This allows user mistakes to pass too silently in my opinion. I would expect an error in the widget to be shown if I return a date  "9/11/2001", "2019-01-33" or "2019-21-12" etc. Currently it could be that some dates work, but the incorrect just become missing values without any notice.
The documentation should also definitely link to https://en.wikipedia.org/wiki/ISO_8601 since most people don't actually use the iso standard for dates and probably don't know what they should look like.

Reply:
+1

Reply:
Not as a part of this PR, but it would be soooo nice if Orange worked with more than one datetime format. Especially when importing from Excel, which pretends to be smart and formats dates in its own way, this would make it a lot easier for people to import datetimes. I assume we can make a simple function that checks the most likely format and applies it to the datetime column... Or actually converts it into ISO, because I suppose nothing would work in the Timeseries add-on otherwise.

Reply:
I forget why I wanted to catch exceptions here. Removed.

Accepting different date formats indeed belongs to a separate issue.